Performance.md

```diff
@@ -0,0 +1,290 @@
+Additionally, it's easy to forget to add new files to your `tsconfig.json`, which means that you might end up with strange editor behavior where those new files are incorrectly analyzed.
+All this can be cumbersome.
+
+`include`/`exclude` help avoid needing to specify these files, but at a cost: files must be discovered by walking through included directories.
+When running through a *lot* of folders, this can slow compilations down.
+Additionally, sometimes a compilation will include lots of unnecessary `.d.ts` files and test files, which can increase compilation time and memory overhead.
+Finally, while `exclude` has some reasonable defaults, certain configurations like mono-repos mean that a "heavy" folders like `node_modules` can still end up being included.
+
+For best practices, we recommend the following:
+
+* Specify only source folders in your project.
+* Don't mix source files from other projects in the same folder.
+* If keeping tests in the same folder as other source files, give them a distinct name so they can easily be excluded.
+* Avoid large build artifacts and dependency folders like `node_modules` in source directories
+
+Here is a reasonable `tsconfig.json` that demonstrates this in action.
+
+```json5
+{
+    "compilerOptions": {
+        // ...
+    },
+    "include": ["src"],
+    "exclude": ["**/node_modules", "**/.*/"],
+}
+```
+
+## Incremental Project Emit
+
+The `--incremental` flag allows TypeScript to save state from the last compilation to a `.tsbuildinfo` file.
+This file is used to figure out the smallest set of files that might to be re-checked/re-emitted since it last ran, much like how TypeScript's `--watch` mode works.
+
+Incremental compiles are enabled by default when using the `composite` flag for project references, but can bring the same speed-ups for any project that opts in.
+
+## Skipping `.d.ts` Checking
+
+By default, TypeScript performs a full re-check of all `.d.ts` files in a project to find issue and inconsistencies; however, this is typically unnecessary.
+Most of the time, the `.d.ts` files are known to already work - the way that types extend each other was already verified once, and declarations that matter will be checked anyway.
+
+TypeScript provides the option to skip type-checking of the `.d.ts` files that it ships with (e.g. `lib.d.ts`) using the `skipDefaultLibCheck` flag.
+
+Alternatively, you can also enable the `skipLibCheck` flag to skip checking *all* `.d.ts` files in a compilation.
+
+These two options can often hide misconfiguration and conflicts in `.d.ts` files, so we suggest using them sparingly.
+
+# Configuring Other Build Tools
+
+TypeScript compilation is often performed with other build tools in mind - especially when writing web apps that might involve a bundler.
+While we can only make suggestions for a few build tools, ideally these techniques can be generalized.
+
+## Concurrent Type-Checking
```

Comment:
I'm not sure I understand the purpose of this paragraph.

Reply:
If you have slow type-checks that impact watch mode, you you might benefit from running type-checking concurrently.

Reply:
So the scenario is that you're using watch mode to keep your output directory up-to-date, rather than to find errors in your code?  And this paragraph says you can have both using a plugin?

Reply:
Yes